presto-main/src/main/java/io/prestosql/connector/system/jdbc/ColumnJdbcTable.java

```diff
@@ -111,24 +139,155 @@ public ConnectorTableMetadata getTableMetadata()
         return METADATA;
     }
 
+    @Override
+    public TupleDomain<ColumnHandle> applyFilter(ConnectorSession connectorSession, Constraint constraint)
+    {
+        TupleDomain<ColumnHandle> tupleDomain = constraint.getSummary();
+        if (tupleDomain.isNone() || !constraint.predicate().isPresent()) {
+            return tupleDomain;
+        }
+        Predicate<Map<ColumnHandle, NullableValue>> predicate = constraint.predicate().get();
+        Set<ColumnHandle> predicateColumns = constraint.getColumns().orElseThrow(() -> new VerifyException("columns not present for a predicate"));
+
+        boolean hasSchemaPredicate = predicateColumns.contains(TABLE_SCHEMA_COLUMN);
+        boolean hasTablePredicate = predicateColumns.contains(TABLE_NAME_COLUMN);
+        if (!hasSchemaPredicate && !hasTablePredicate) {
+            // No filter on schema name and table name at all.
+            return tupleDomain;
+        }
+
+        Session session = ((FullConnectorSession) connectorSession).getSession();
+
+        Optional<String> catalogFilter = tryGetSingleVarcharValue(tupleDomain, TABLE_CATALOG_COLUMN);
+        Optional<String> schemaFilter = tryGetSingleVarcharValue(tupleDomain, TABLE_SCHEMA_COLUMN);
+        Optional<String> tableFilter = tryGetSingleVarcharValue(tupleDomain, TABLE_NAME_COLUMN);
+
+        if (catalogFilter.equals(Optional.of("")) ||
+                schemaFilter.equals(Optional.of("")) ||
+                tableFilter.equals(Optional.of(""))) {
+            return TupleDomain.none();
+        }
+
+        List<String> catalogs = listCatalogs(session, metadata, accessControl, catalogFilter).keySet().stream()
+                .filter(catalogName -> predicate.test(ImmutableMap.of(TABLE_CATALOG_COLUMN, toNullableValue(catalogName))))
+                .collect(toImmutableList());
+
+        List<CatalogSchemaName> schemas = catalogs.stream()
+                .flatMap(catalogName ->
+                        listSchemas(session, metadata, accessControl, catalogName, schemaFilter).stream()
+                                .filter(schemaName -> !hasSchemaPredicate || predicate.test(ImmutableMap.of(
+                                        TABLE_CATALOG_COLUMN, toNullableValue(catalogName),
+                                        TABLE_SCHEMA_COLUMN, toNullableValue(schemaName))))
+                                .map(schemaName -> new CatalogSchemaName(catalogName, schemaName)))
+                .collect(toImmutableList());
+
+        if (!hasTablePredicate) {
+            return TupleDomain.withColumnDomains(ImmutableMap.of(
+                    TABLE_CATALOG_COLUMN, schemas.stream()
```

Comment:
Using `ImmutableMap.builder()` is less sensitive to reformatting changes that may end up placing arguments in separate lines (and even if that happens, it's still clear which ones are keys and which ones are values). TLDR, when there are multiple entries, it's more robust to use the builder form than the multi-entry `ImmutableMap.of()`.